src/docmanager/index.ts

```diff
@@ -0,0 +1,514 @@
+   * The list of kernel specs associated with the running sessions.
+   */
+  specs: IKernelSpecId[];
+}
+
+
+/**
+ * The document context object.
+ */
+export interface IDocumentContext {
+  /**
+   * A signal emitted when the kernel changes.
+   */
+  kernelChanged: ISignal<IDocumentContext, IKernel>;
+
+  /**
+   * A signal emitted when the path changes.
+   */
+  pathChanged: ISignal<IDocumentContext, string>;
+
+  /**
+   * A signal emitted when the model is saved or reverted.
+   */
+  dirtyCleared: ISignal<IDocumentContext, void>;
+
+  /**
+   * The current kernel associated with the document.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  kernel: IKernel;
+
+  /**
+   * The current path associated with the document.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  path: string;
+
+  /**
+   * Change the current kernel associated with the document.
+   */
+  changeKernel(options: IKernelId): Promise<IKernel>;
+
+  /**
+   * Save the document contents to disk.
+   */
+  save(): Promise<void>;
+
+  /**
+   * Revert the document contents to disk contents.
+   */
+  revert(): Promise<void>;
+
+  /**
+   * Get the list of running sessions.
+   */
+  listRunningSessions(): Promise<ISessionInfo>;
+
+  /**
+   * Add a sibling widget to the document manager.
+   *
+   * @param widget - The widget to add to the document manager.
+   *
+   * @returns A disposable used to remove the sibling if desired.
+   */
+  addSibling(widget: Widget): IDisposable;
+}
+
+
+/**
+ * The interface for a widget factory.
+ */
+export
+interface IWidgetFactory<T extends Widget> {
+  /**
+   * Create a new widget given a document model and a context.
+   */
+  createNew(model: IDocumentModel, context: IDocumentContext): T;
+
+  /**
+   * Get the default kernel name given a path and a list of specs.
+   */
+  getDefaultKernel(path: string, specs: IKernelSpecId[]): string;
+
+  /**
+   * Get the preferred kernel names given a path and a list of specs.
+   */
+  getPreferredKernels(path: string, specs: IKernelSpecId[]): string[];
+
+  /**
+   * Get the preferred widget title given a path.
+   */
+  getWidgetTitle(path: string): string;
+}
+
+
+/**
+ * The options uses to register a widget factory.
+ */
+export
+interface IWidgetFactoryOptions {
+  /**
+   * The type of file the widget can view.
+   */
+  fileType: string;
+
+  /**
+   * The name of the widget to display in dialogs.
+   */
+  displayedName: string;
+
+  /**
+   The object used to create new widgets.
+   */
+  widgetFactory: IWidgetFactory<Widget>;
+
+  /**
+   * The registered name of the model type used to create the widgets.
+   */
+  modelName: string;
+}
+
+
+/**
+ * The interface for a model factory.
+ */
+export
+interface IModelFactory {
+  /**
+   * Create a new model.
+   */
+  createNew(path: string, kernelSpecs: IKernelSpecId[]): IDocumentModel;
+}
+
+
+/**
+ * The options used to register a model factory.
+ */
+export
+interface IModelFactoryOptions {
```

Comment:
Oh, I get it. Since we can't have static functions on interfaces, you wrap the IModelFactory interface in this so we effectively have static functions on the IModelFactory, but specified in an interface.